src/darwin/Framework/CHIP/MTREndpointInfo.mm

```diff
@@ -0,0 +1,210 @@
+/**
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MTREndpointInfo_Internal.h"
+
+#import "MTRAttributeTLVValueDecoder_Internal.h"
+#import "MTRDeviceTypeRevision.h"
+#import "MTRLogging_Internal.h"
+#import "MTRStructsObjc.h"
+
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+
+#include <deque>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+
+enum class EndpointMark : uint8_t {
+    NotVisited = 0,
+    Visiting,
+    Visited,
+    ParentAssigned = NotVisited, // != Visited
+};
+
+MTR_DIRECT_MEMBERS
+@implementation MTREndpointInfo {
+    EndpointId _endpointID;
+    EndpointMark _mark; // used by populateChildrenForEndpoints:
+}
+
+- (instancetype)initWithEndpointID:(NSNumber *)endpointID
+                       deviceTypes:(NSArray<MTRDeviceTypeRevision *> *)deviceTypes
+                         partsList:(NSArray<NSNumber *> *)partsList
+{
+    self = [super init];
+    _endpointID = endpointID.unsignedShortValue;
+    _deviceTypes = [deviceTypes copy];
+    _partsList = [partsList copy];
+    _children = @[];
+    return self;
+}
+
+- (NSNumber *)endpointID
+{
+    return @(_endpointID);
+}
+
+- (NSString *)description
+{
+    return [NSString stringWithFormat:@"<%@ %u>", self.class, _endpointID];
+}
+
++ (BOOL)populateChildrenForEndpoints:(NSDictionary<NSNumber *, MTREndpointInfo *> *)endpoints
+{
+    // Populate the child list of each endpoint, ensuring no cycles (these are disallowed
+    // by the spec, but we need to ensure we don't create a retain cycle even under invalid
+    // input). Conservatively assume all endpoints use the Full-Family Pattern.
+    // Refer to "Endpoint Composition" in the Matter specification for details.
+    MTREndpointInfo * root = endpoints[@0];
+    if (root == nil) {
+        MTR_LOG_ERROR("Missing root endpoint, not populating endpoint hierarchy");
+        return NO;
+    }
+
+    // Perform a depth-first search with an explicit stack and create a list of endpoint
+    // IDs in reverse topological order. The _parentID field is used mark traversal states.
```

Comment:
What _parentID field?  Also, "to mark"?

Reply:
Ah, I had one at some point, will fix the comment